compliance_checker/ioos.py

```diff
@@ -346,6 +348,353 @@ def check_units(self, ds):
+
+        # check for component, disciminant
+        for instr in instr_vars:
+            if instr in ds.variables:
+                compnt = getattr(ds.variables[instr], "component", None)
+                m = [f"component attribute of {instr} ({compnt}) must be a string"]
+                if compnt:
+                    results.append(
+                        Result(BaseCheck.MEDIUM, isinstance(compnt, str), "instrument_variable", m)
+                    )
+                else:
+                    results.append(Result(BaseCheck.MEDIUM, True, "instrument_variable", m))
+
+                disct = getattr(ds.variables[instr], "discriminant", None)
+                m = [f"discriminant attribute of {instr} ({disct}) must be a string"]
+                if disct:
+                    results.append(
+                        Result(BaseCheck.MEDIUM, isinstance(disct, str), "instrument_variable", m)
+                    )
+                else:
+                    results.append(Result(BaseCheck.MEDIUM, True, "instrument_variable", m))
+
+        return results
+
+    def check_qartod_variables_references(self, ds):
+        """
+        For any variables that are deemed QARTOD variables, check that they
+        contain the "references" attribute and that the value of the attribute
+        is a valid URL.
+
+        Args:
+            ds (netCDF4.Dataset): open Dataset
+
+        Returns:
+            list of Results
+        """
+
+        vldr = validators.Validator()
+        vldr.require_presence_of("scheme", "host")
+        vldr.allow_schemes("http", "https")
+
+        results = []
+        ctxt = "qartod_variable:references"
+        for v in ds.get_variables_by_attributes(standard_name=lambda x: x in self._qartod_std_names):
+            msg = f"\"references\" attribute for variable \"{v.name}\" must be a valid URL"
+            val = True
+            ref = getattr(v, "references", b'')
+            url = api.uri_reference(ref)
+
+            try:
+                vldr.validate(url)
+            except exceptions.MissingComponentError:
+                val = False
+            results.append(Result(BaseCheck.MEDIUM, val, ctxt, [msg]))
+
+        return results
+
+    def check_wmo_platform_code(self, ds):
+        """
+        If a WMO Platform Code is given as a global variable, check that it is
+        well-formed. According to the WMO, valid codes are a numeric string
+        comprised of 5 or 7 characters.
+
+        Args:
+            ds (netCDF4.Dataset): open Dataset
+
+        Returns:
+            Result
+        """
+
+        valid = True
+        ctxt = "wmo_platform_code"
+        msg = "The wmo_platform_code must be a numeric string of 5 or 7 characters"
+
+        code = getattr(ds, "wmo_platform_code", None)
+        if code:
+           if not (
+               isinstance(code, str) and
+               code.isnumeric() and
+               (len(code) == 5 or len(code) == 7)
```

Comment:
Don't call `len` twice here.

Reply:
Algorithmic complexity of `(len(str))`. Still, I made it better.